File: src/main/scala/s/template_method/delegate/StringDisplayDelegate.scala
package s.template_method.delegate

final class StringDisplayDelegate(private val string: String) extends DisplayDelegate {
  private val width = string.length

  override def open(): Unit = {
    printLine()
  }

  override def print(): Unit = {
    Console.println("|" + string + "|")
  }

  override def close(): Unit = {
    printLine()
  }

  private def printLine(): Unit = {
    Console.print("+")
    for (_ <- 0 until width) {
      Console.print("-")
    }
    Console.println("+")
  }
}

File: src/main/kotlin/k/template_method/delegate/StringDisplayDelegate.kt
package k.template_method.delegate

class StringDisplayDelegate(private val string: String) : DisplayDelegate {
  private val width: Int = string.length

  override fun open() {
    printLine()
  }

  override fun print() {
    println("|$string|")
  }

  override fun close() {
    printLine()
  }

  private fun printLine() {
    print("+")
    for (i in 0 until width) {
      print("-")
    }
    println("+")
  }
}
